In [2]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})


In [3]:
df


name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

In [4]:
from pydataset import data

In [15]:
df['passing_english'] = df.reading > 69

In [16]:
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67            False
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72             True

In [17]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
1      Jane    88       79       67            False
0     Sally    62       85       80             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72             True

Sort the english grades first by passing_english and then by student name.
All the students that are failing english should be first, and within the 
students that are failing english they should be ordered alphabetically. 
The same should be true for the students passing english.
(Hint: you can pass a list to the .sort_values method)

In [20]:
df.sort_values(['passing_english','name'], ascending=[True, True])

name  math  english  reading  passing_english
1      Jane    88       79       67            False
4       Ada    77       92       98             True
11     Alan    92       62       72             True
8    Albert    92       62       87             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
5      John    79       76       93             True
7     Marie    93       63       90             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81             True

Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [22]:
df.sort_values(['passing_english','english'], ascending=[True,False])

name  math  english  reading  passing_english
1      Jane    88       79       67            False
10    Isaac    92       99       93             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
0     Sally    62       85       80             True
9   Richard    69       80       94             True
5      John    79       76       93             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
11     Alan    92       62       72             True

Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [25]:
df['overall_grade'] = ((df.math + df.english + df.reading) / 3).astype(int)

In [26]:
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True             75
1      Jane    88       79       67            False             78
2     Suzie    94       74       95             True             87
3     Billy    98       96       88             True             94
4       Ada    77       92       98             True             89
5      John    79       76       93             True             82
6    Thomas    82       64       81             True             75
7     Marie    93       63       90             True             82
8    Albert    92       62       87             True             80
9   Richard    69       80       94             True             81
10    Isaac    92       99       93             True             94
11     Alan    92       62       72             True             75

In [27]:
mpg = data('mpg')

In [28]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [29]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


How many rows and columns are there?

What are the data types of each column?


-234 rows and 11 columns
-object,object,float,int,int,object,object,int,int,object,object

In [30]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

Rename the cty column to city.

In [141]:
mpg.rename(columns={'cty': 'city'})


manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  mileage_differnce  mileage_difference  average_mileage  
1    compact                 11                  11             23.5  
2    compact                  8                   8             25.0  
3    compact                 11                  11             25.5  
4    compact                  9                   9             25.5  
5    compact                 10                  10             21.0  
..       ...                ...                 ...              ...  
230  midsize                  9                   9             23.5  
231  midsize                  8                   8             25.0  
232  midsize                 10                  10             21.0  
233  midsize                  8                   8             22.0  
234  midsize                  9                   9             21.5  

[234 rows x 14 columns]

Rename the hwy column to highway.


In [39]:
mpg.rename(columns={'hwy': 'highway'})

manufacturer   model  displ  year  cyl       trans drv  cty  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16       26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

Do any cars have better city mileage than highway mileage?


In [143]:
mpg[mpg.city > mpg.hwy]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, hwy, fl, class, mileage_differnce, mileage_difference, average_mileage]
Index: []

Create a column named mileage_difference this column 
should contain the difference between highway and city mileage for each 
car.


In [46]:
mpg['mileage_difference'] = mpg.hwy - mpg.cty

In [47]:
mpg


manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_differnce  mileage_difference  
1    compact                 11                  11  
2    compact                  8                   8  
3    compact                 11                  11  
4    compact                  9                   9  
5    compact                 10                  10  
..       ...                ...                 ...  
230  midsize                  9                   9  
231  midsize                  8                   8  
232  midsize                 10                  10  
233  midsize                  8                   8  
234  midsize                  9                   9  

[234 rows x 13 columns]

Which car (or cars) has the highest mileage difference?


In [56]:
highest = mpg.mileage_difference.max()

In [58]:
mpg[mpg.mileage_difference == highest]

manufacturer       model  displ  year  cyl     trans drv  cty  hwy fl  \
107        honda       civic    1.8  2008    4  auto(l5)   f   24   36  c   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f   29   41  d   

          class  mileage_differnce  mileage_difference  
107  subcompact                 12                  12  
223  subcompact                 12                  12

Which compact class car has the lowest highway mileage? The best?


In [138]:
compact = mpg[mpg['class'] == 'compact']

In [139]:
compact.hwy.min()

23

Create a column named average_mileage that is the mean of the city and highway mileage.

In [80]:
mpg['average_mileage'] = (mpg.cty + mpg.hwy) / 2

Which dodge car has the best average mileage? The worst?


In [125]:
mpg.average_mileage.max()
mpg[mpg.average_mileage == 39.5]

manufacturer       model  displ  year  cyl       trans drv  cty  hwy fl  \
222   volkswagen  new beetle    1.9  1999    4  manual(m5)   f   35   44  d   

          class  mileage_differnce  mileage_difference  average_mileage  
222  subcompact                  9                   9             39.5

mpg.average_mileage.min()

Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

In [ ]:
mpg.average_mileage.min()

In [127]:
mpg[mpg.average_mileage == 10.5]

manufacturer                model  displ  year  cyl       trans drv  cty  \
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4    9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4    9   

     hwy fl   class  mileage_differnce  mileage_difference  average_mileage  
55    12  e  pickup                  3                   3             10.5  
60    12  e     suv                  3                   3             10.5  
66    12  e  pickup                  3                   3             10.5  
70    12  e  pickup                  3                   3             10.5  
127   12  e     suv                  3                   3             10.5

In [95]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62 entries, Arctic fox to Red fox
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   body    62 non-null     float64
 1   brain   62 non-null     float64
dtypes: float64(2)
memory usage: 1.5+ KB


In [99]:
mammals = data('Mammals')

In [111]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


How many rows and columns are there?
4 and 107
What are the data types of each column?
float, float, bool, bool

What is the the weight of the fastest animal?


In [104]:
 mammals.speed.max()

In [109]:
mammals[mammals['speed'] == 110.0]

weight  speed  hoppers  specials
53    55.0  110.0    False     False

What is the overal percentage of specials?


In [113]:
sum(mammals['specials'] == True) / len(mammals)

0.09345794392523364

How many animals are hoppers that are above the median speed? What percentage is this?

In [145]:
mammals['median_speed'] = sum(mammals['speed']) / len(mammals)

In [115]:
mammals

weight  speed  hoppers  specials  median_speed
1    6000.000   35.0    False     False     46.208411
2    4000.000   26.0    False     False     46.208411
3    3000.000   25.0    False     False     46.208411
4    1400.000   45.0    False     False     46.208411
5     400.000   70.0    False     False     46.208411
6     350.000   70.0    False     False     46.208411
7     300.000   64.0    False     False     46.208411
8     260.000   70.0    False     False     46.208411
9     250.000   40.0    False     False     46.208411
10   3800.000   25.0    False      True     46.208411
11   1000.000   60.0    False     False     46.208411
12    900.000   70.0    False     False     46.208411
13    900.000   56.0    False     False     46.208411
14    800.000   29.0    False     False     46.208411
15    750.000   57.0    False     False     46.208411
16    500.000   32.0    False     False     46.208411
17    450.000   56.0    False     False     46.208411
18    300.000   72.0    False     False     46.208411
19    300.000   90.0    False     False     46.208411
20    250.000   80.0    False     False     46.208411
21    250.000   56.0    False     False     46.208411
22    170.000   80.0    False     False     46.208411
23    150.000   48.0    False     False     46.208411
24    130.000   70.0    False     False     46.208411
25    120.000   80.0    False     False     46.208411
26    120.000   61.0    False     False     46.208411
27    110.000   33.0    False     False     46.208411
28    100.000   64.0    False     False     46.208411
29     85.000   55.0    False     False     46.208411
30     80.000   65.0    False     False     46.208411
31     72.000   56.0    False     False     46.208411
32     70.000   45.0    False     False     46.208411
33     65.000   60.0    False     False     46.208411
34     62.000   81.0    False     False     46.208411
35     50.000  100.0    False     False     46.208411
36     50.000   60.0    False     False     46.208411
37     50.000   40.0    False     False     46.208411
38     50.000   47.0    False     False     46.208411
39     37.000  105.0    False     False     46.208411
40     35.000   80.0    False     False     46.208411
41     34.000   97.0    False     False     46.208411
42     30.000   97.0    False     False     46.208411
43     30.000   80.0    False     False     46.208411
44     30.000   45.0    False     False     46.208411
45     20.000   81.0    False     False     46.208411
46    400.000   40.0    False     False     46.208411
47    300.000   48.0    False     False     46.208411
48    230.000   56.0    False     False     46.208411
49    150.000   59.0    False     False     46.208411
50    135.000   48.0    False     False     46.208411
51     65.000   65.0    False     False     46.208411
52     60.000   60.0    False     False     46.208411
53     55.000  110.0    False     False     46.208411
54     45.000   50.0    False     False     46.208411
55     40.000   64.0    False     False     46.208411
56     25.000   67.0    False     False     46.208411
57     20.000   70.0    False     False     46.208411
58     16.000   65.0    False     False     46.208411
59     12.000   24.0    False      True     46.208411
60     11.000   30.0    False      True     46.208411
61     10.000   56.0    False     False     46.208411
62      7.000   60.0    False     False     46.208411
63      6.000   72.0    False     False     46.208411
64      5.000   64.0    False     False     46.208411
65      5.000   27.0    False      True     46.208411
66      3.000   16.0    False      True     46.208411
67    127.000   32.0    False     False     46.208411
68     70.000   40.0    False      True     46.208411
69     13.000   37.0    False      True     46.208411
70      9.000    3.2    False      True     46.208411
71      4.000   16.0    False     False     46.208411
72      0.600   36.0    False     False     46.208411
73      0.600   20.0    False     False     46.208411
74      0.5

In [120]:
greater = sum(mammals['speed'] > mammals['median_speed'])

In [121]:
greater

57

In [122]:
greater / len(mammals)

0.5327102803738317